src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,265 @@
+
+        valid_partitions_end = validate_replicas(
+          valid_partitions_end,
+          topic.partitions.end(),
+          std::back_inserter(invalid_partitions),
+          error_code::invalid_replica_assignment,
+          "Empty replica list specified in partition reassignment.",
+          [](const reassignable_partition& partition) {
+              return !partition.replicas->empty();
+          });
+
+        valid_partitions_end = validate_replicas(
+          valid_partitions_end,
+          topic.partitions.end(),
+          std::back_inserter(invalid_partitions),
+          error_code::invalid_replica_assignment,
+          "Duplicate replica ids in partition reassignment replica list",
+          [](const reassignable_partition& partition) {
+              auto replicas_copy = *partition.replicas;
+              std::sort(replicas_copy.begin(), replicas_copy.end());
+              auto no_duplicates_it = std::adjacent_find(
+                replicas_copy.cbegin(), replicas_copy.cend());
+              return no_duplicates_it == replicas_copy.cend();
+          });
+
+        valid_partitions_end = validate_replicas(
+          valid_partitions_end,
+          topic.partitions.end(),
+          std::back_inserter(invalid_partitions),
+          error_code::invalid_replica_assignment,
+          "Invalid broker id in replica list",
+          [](const reassignable_partition& partition) {
+              auto negative_node_id_it = std::find_if(
+                partition.replicas->begin(),
+                partition.replicas->end(),
+                [](const model::node_id& node_id) { return node_id < 0; });
+              return negative_node_id_it == partition.replicas->end();
+          });
+
+        valid_partitions_end = validate_replicas(
+          valid_partitions_end,
+          topic.partitions.end(),
+          std::back_inserter(invalid_partitions),
+          error_code::invalid_replica_assignment,
+          "Replica assignment has brokers that are not alive",
+          [&all_node_ids](const reassignable_partition& partition) {
+              auto unkown_broker_id_it = std::find_if(
+                partition.replicas->begin(),
+                partition.replicas->end(),
+                [all_node_ids](const model::node_id& node_id) {
+                    return std::find(
+                             all_node_ids.begin(), all_node_ids.end(), node_id)
+                           == all_node_ids.end();
+                });
+              return unkown_broker_id_it == partition.replicas->end();
+          });
+
+        topic_response.partitions = std::move(invalid_partitions);
+        resp_it = std::move(topic_response);
+
+        // If there are no valid partitions ...
+        if (valid_partitions_end == topic.partitions.begin()) {
+            // Then remake the partitions vector as empty
+            topic.partitions = std::vector<reassignable_partition>{};
```

Comment:
If the code flowed with always using the valid begin iterator , then it would eventually hit a loop that never iterates because valid_end == end, there would never need to be explicit logic for these cases